File: nahomy-c2-backend/src/domain/entities/Usuario.ts
import { UserId } from '../value-objects/UserId'
import { PasswordHash } from '../value-objects/PasswordHash'

export type UsuarioRole = 'Administrador' | 'Operador' | 'Auditor'

export class Usuario {
  private readonly _id: UserId
  private _nombre: string
  private _passwordHash: PasswordHash
  private _rol: UsuarioRole

  constructor(id: UserId, nombre: string, passwordHash: PasswordHash, rol: UsuarioRole) {
    this._id = id
    this._nombre = nombre
    this._passwordHash = passwordHash
    this._rol = rol
  }

  get id(): UserId {
    return this._id
  }

  get nombre(): string {
    return this._nombre
  }

  get passwordHash(): PasswordHash {
    return this._passwordHash
  }

  get rol(): UsuarioRole {
    return this._rol
  }

  cambiarRol(nuevoRol: UsuarioRole): void {
    this._rol = nuevoRol
  }

  actualizarNombre(nombre: string): void {
    this._nombre = nombre
  }

  actualizarPassword(hash: PasswordHash): void {
    this._passwordHash = hash
  }
}

File: packages/server/src/domain/entities/usuario.py
from __future__ import annotations
from typing import Literal
from ..value_objects import UserId, PasswordHash

UsuarioRole = Literal["Administrador", "Operador", "Auditor"]


class Usuario:
    def __init__(
        self,
        id: UserId,
        nombre: str,
        password_hash: PasswordHash,
        rol: UsuarioRole,
    ):
        self._id = id
        self._nombre = nombre
        self._password_hash = password_hash
        self._rol = rol

    @property
    def id(self) -> UserId:
        return self._id

    @property
    def nombre(self) -> str:
        return self._nombre

    @property
    def password_hash(self) -> PasswordHash:
        return self._password_hash

    @property
    def rol(self) -> UsuarioRole:
        return self._rol

    def cambiar_rol(self, nuevo_rol: UsuarioRole) -> None:
        self._rol = nuevo_rol

    def actualizar_nombre(self, nombre: str) -> None:
        self._nombre = nombre

    def actualizar_password(self, hash_value: PasswordHash) -> None:
        self._password_hash = hash_value
